### ISSUE TITLE
[MPU9250] incorrect yaw value

### ISSUE THREAD

**Issue by thorikawa** [m5stack/M5Stack#130]
I recently purchased M5Stack Gray and tried [MPU9250BasicAHRS.ino](https://github.com/m5stack/M5Stack/blob/master/examples/Modules/MPU9250/MPU9250BasicAHRS/MPU9250BasicAHRS.ino). When I run the sketch and rotate the device, pitch and roll values look correct, but yaw does not. Yaw value keeps almost same value even if I am rotating it. How can I get the correct yaw value?

I posted the video here: https://www.youtube.com/watch?v=tkF0MKzGXqA

Also pasted first a few lines of the logs:
```
OK
MPU9250 I AM 71 I should be 71
MPU9250 is online...
x-axis self test: acceleration trim within : 2.6% of factory value
y-axis self test: acceleration trim within : -3.0% of factory value
z-axis self test: acceleration trim within : -1.4% of factory value
x-axis self test: gyration trim within : 0.4% of factory value
y-axis self test: gyration trim within : -0.1% of factory value
z-axis self test: gyration trim within : -0.6% of factory value
MPU9250 initialized for active data mode....
AK8963 I AM 48 I should be 48
AK8963 initialized for active data mode....
X-Axis sensitivity adjustment value 1.19
Y-Axis sensitivity adjustment value 1.20
Z-Axis sensitivity adjustment value 1.15
ax = -127.81 ay = 0.12 az = 1000.00 mg
gx = 0.07 gy = 0.06 gz = -0.94 deg/s
mx = -6677 my = 4283 mz = -3901 mG
q0 = 0.03 qx = 0.44 qy = -0.07 qz = 0.89
Yaw, Pitch, Roll: -188.25, -52.50, -9.47
rate = 0.13 Hz

ax = -130.13 ay = -2.26 az = 996.64 mg
gx = 0.32 gy = 0.05 gz = -1.11 deg/s
mx = -6670 my = 4265 mz = -3911 mG
q0 = 0.58 qx = -0.00 qy = -0.04 qz = 0.81
Yaw, Pitch, Roll: 99.98, -2.05, -3.67
rate = 754.01 Hz

ax = -129.76 ay = -2.69 az = 999.94 mg
gx = 0.15 gy = 0.05 gz = -1.18 deg/s
mx = -6661 my = 4256 mz = -3902 mG
q0 = 0.78 qx = -0.09 qy = 0.07 qz = 0.62
Yaw, Pitch, Roll: 68.41, 13.09, -2.86
rate = 756.43 Hz

ax = -129.76 ay = -1.77 az = 998.05 mg
gx = 0.31 gy = 0.03 gz = -0.98 deg/s
mx = -6659 my = 4262 mz = -3901 mG
q0 = 0.82 qx = -0.07 qy = 0.07 qz = 0.56
Yaw, Pitch, Roll: 59.45, 11.09, -1.66
rate = 751.43 Hz

ax = -129.70 ay = -0.67 az = 996.34 mg
gx = 0.20 gy = 0.11 gz = -1.11 deg/s
mx = -6675 my = 4270 mz = -3899 mG
q0 = 0.84 qx = -0.05 qy = 0.06 qz = 0.53
Yaw, Pitch, Roll: 55.65, 9.19, -0.86
rate = 750.63 Hz
```

Any help would be appreciated.

**Comment by sakabin** [comment]
Sorry, we also found this problem and I am trying to solve it

**Comment by thorikawa** [comment]
I figured it out myself. First of all, the current magnetometer is very sensitive, so it is affected by various conditions including whether USB is inserted or not, whether PC is located nearby the device, etc. Currently magnetometer bias is hard-coded in the [sketch](https://github.com/m5stack/M5Stack/blob/master/examples/Modules/MPU9250/MPU9250BasicAHRS/MPU9250BasicAHRS.ino#L192), but it should be dynamically adjusted because it highly depends on the individual condition.

Also, to make matters worse, M5Stack Gray has powerful neodymium magnets inside the bottom module which also strongly affect the magnetometer.

What I did to solve this issue is that
1. Remove neodymium magnets from the bottom module.
2. Calibrate the magnetometer according to [this article](https://appelsiini.net/2018/calibrate-magnetometer/) and change the hard-coded value.
Also I pass -IMU.mz insted of IMU.mz to [MahonyQuaternionUpdate](https://github.com/m5stack/M5Stack/blob/master/examples/Modules/MPU9250/MPU9250BasicAHRS/MPU9250BasicAHRS.ino#L224) since according to the [document](http://www.invensense.com/wp-content/uploads/2015/02/PS-MPU-9250A-01-v1.1.pdf), MPU9250 magnetometer's z axis is inverse of other sensor's z axis.

I don't send pull request, because as I already said, the bias value depends on the individual device and environment. I think the sketch might as well include dynamic calibration logic and neodymium magnets should be removed especially for this device.

Thank you.
